packages/amesos2/src/Amesos2_Tacho_def.hpp

```diff
@@ -0,0 +1,371 @@
+TachoSolver<Matrix,Vector>::numericFactorization_impl()
+{
+  int status = 0;
+
+  if ( this->root_ ) {
+    value_type_array values;
+
+#ifndef HAVE_TEUCHOS_COMPLEX
+    if(single_process_optim_check()) {
+      // in the optimized case we read the values directly from the matrix
+      // without converting through the Teuchos::Array setup.
+      auto sp_values = this->matrixA_->returnValues();
+        TEUCHOS_TEST_FOR_EXCEPTION(sp_values == nullptr,
+          std::runtime_error, "Amesos2 Runtime Error: sp_values returned null");
+      values = value_type_array(sp_values, this->globalNumNonZeros_);
+    }
+    else
+#endif
+    {
+      // Non optimized case used the arrays set up in loadA_impl
+      values = value_type_array(this->nzvals_.getRawPtr(), this->globalNumNonZeros_);
+    }
+
+    data_.solver.factorize(values);
+  }
+
+  return status;
+}
+
+template <class Matrix, class Vector>
+int
+TachoSolver<Matrix,Vector>::solve_impl(const Teuchos::Ptr<MultiVecAdapter<Vector> >       X,
+                                   const Teuchos::Ptr<const MultiVecAdapter<Vector> > B) const
+{
+  using Teuchos::as;
+
+  const global_size_type ld_rhs = this->root_ ? X->getGlobalLength() : 0;
+  const size_t nrhs = X->getGlobalNumVectors();
+
+  const size_t val_store_size = as<size_t>(ld_rhs * nrhs);
+  Teuchos::Array<tacho_type> xValues(val_store_size);
+  Teuchos::Array<tacho_type> bValues(val_store_size);
+
+  {                             // Get values from RHS B
+#ifdef HAVE_AMESOS2_TIMERS
+    Teuchos::TimeMonitor mvConvTimer(this->timers_.vecConvTime_);
+    Teuchos::TimeMonitor redistTimer(this->timers_.vecRedistTime_);
+#endif
+    Util::get_1d_copy_helper<MultiVecAdapter<Vector>,
+                             tacho_type>::do_get(B, bValues(),
+                                               as<size_t>(ld_rhs),
+                                               ROOTED, this->rowIndexBase_);
+  }
+
+  int ierr = 0; // returned error code
+
+  if ( this->root_ ) {
+    {                           // Do solve!
+  #ifdef HAVE_AMESOS2_TIMER
+      Teuchos::TimeMonitor solveTimer(this->timers_.solveTime_);
+  #endif
+      // validate
+      int i_ld_rhs = as<int>(ld_rhs);
+
+      for(size_t j = 0 ; j < nrhs; j++) {
+        // TODO: Decide how to organize this best for Amesos2
+        // Clarify behavior for Cuda - exception? configure error?
+#ifdef KOKKOS_HAVE_OPENMP
+        typedef Kokkos::OpenMP                                  DeviceSpaceType;
+#else
+        typedef Kokkos::Serial                                  DeviceSpaceType;
+#endif
+        typedef typename Tacho::Solver<tacho_type,DeviceSpaceType>::
+          value_type_matrix solve_array_t;
+
+        const int n = 1;
+        solve_array_t x(&xValues.getRawPtr()[j*i_ld_rhs], this->globalNumRows_, n);
+        solve_array_t b(&bValues.getRawPtr()[j*i_ld_rhs], this->globalNumRows_, n);
+        solve_array_t t("t", this->globalNumRows_, n);
+        data_.solver.solve(x, b, t);
+
```

Comment:
Consider that this solve routine is called many times. x and b are just wrapping pointers and it is okay. Constructing the temporary workspace of "t" is not necessary to be called multiple times if it is created before. Can you put this as member variable and construct the array when it is needed ?

Reply:
`rowptr`'s entry type is actually a function of the Kokkos execution space's default `size_type`.

Reply:
@mhoemmen 

I think that this comment is about my previous comment. 

In my code, I do not inherit oridnal_type and size_type from Kokkos execution space but they are simply "int" and "size_t" unless cmake configuration enforces to use int as its size_type (SIERRA uses int only).

@MicheldeMessieres 

Do we have full unit tests that cover all ordinal type combinations ?

Reply:
Currently Amesos2 is not testing both Tacho_ENABLE_INT_INT ON/OFF. In the current setup this could be tested by manually setting ON/OFF for SEMS. If we wanted to have Amesos2 run two different tests for both cases we might want to rethink how this is setup since it is currently controlled at the Tacho level and coded into ShyLUTacho_config.h. I'm not sure what the right design is to avoid confusion what the config options mean and where we would want Amesos2 controlling this.

Reply:
I've pushed some new changes which move the workspace t to be allocated just once. Also some other changes to remove some typedefs that were not needed. This was checked on SEMS to the latest develop for complex on/off.

Reply:
It is okay. Tacho_ENABLE_INT_INT is for seirra and they do not use tacho through Amesos2. By default, we test Tacho_ENABLE_INT_INT=OFF. if this is not configured in cmake, tacho size_type is always size_t.